Update permissions computation to account for permissionsHeld membrane requirements.

dwrensha opened the pull request:
This is a continuation of https://github.com/sandstorm-io/sandstorm/pull/1620, which accidentally got closed.

I've implemented a selective forward-chaining algorithm that avoids the performance issues of the previous implementation.

dwrensha commented:
Garply, retest this please.

dwrensha commented:
Garply, retest this please.

jparyani commented:
Garply, retest this please

kentonv commented on shell/packages/sandstorm-permissions/permissions.js:
```diff
@@ -291,100 +252,173 @@ class RequirementSet {
+  // Our permissions computation can be framed as a propositional HORNSAT problem; this `Variable`
+  // class represents a variable in that sense. There is a variable for every (grain ID, vertex ID,
+  // permission ID) triple. In any given computation, we only explicitly construct those
+  // variables that know might actually be relevant to the result.
```
that know -> that we know

kentonv commented on shell/packages/sandstorm-permissions/permissions.js:
```diff
@@ -291,100 +252,173 @@ class RequirementSet {
-class ConditionalPermissionSet {
-  // A PermissionSet that is contingent upon some requirements.
+  requirementsAreMet() {
+    return this.numUnmetRequirements <= 0;
```
Presumably a negative value here suggests a horrible bug where some requirements may be being ignored? Maybe `decrementRequirements()` should throw if the count goes negative?

kentonv commented on shell/packages/sandstorm-permissions/permissions.js:
```diff
@@ -291,100 +252,173 @@ class RequirementSet {
+  }
+}
+
+const permissionIdPattern = Match.OneOf({ canAccess: null }, { appDefined: Match.Integer });
```
This feels bulky to me. Possible alternative: Match.OneOf(Number, "canAccess"), i.e. the value is either a number or the special string value "canAccess". You can then represent a permission set as an object whose keys are numbers or "canAccess". This actually makes more sense than an array in that a permission array is really an infinite array with an all-false tail, so now we have an infinite object where unset fields are considered false...

dwrensha replied:
I've switched to a plain object. Now a permission ID is a stringified integer >= -1, where "-1" means "can access".  This does simplify a few things!

kentonv commented on shell/packages/sandstorm-permissions/permissions.js:
```diff
@@ -291,100 +252,173 @@ class RequirementSet {
-    result.requirements.addMembraneRequirements(token.requirements);
-    result.tokensUsed[token._id] = true;
-    return result;
+  directIsMet(permissionId) {
```
These methods could use doc comments.

dwrensha commented:
Garply, retest this please.

dwrensha commented:
I've addressed the comments that have come in so far.

I don't know what the deal is with the integration tests... Garply is acting strangely.

jparyani commented:
Garply, retest this please.

jparyani commented:
That was my bad. I was resizing the disk and had to reboot the machine.

kentonv commented on shell/packages/sandstorm-permissions/permissions.js:
```diff
@@ -256,18 +256,17 @@ class RequirementSet {
-// the permissions enumerated in a grain's `ViewInfo`, but also the usually-implicit "can
-// access the grain at all" permission. A permission ID, defined here, allows us to refer
-// to either of these types of permissions.
+// A permission ID is a stringified integer greater than or equal to -1. The permission "-1"
```
Why stringified?

Why "-1" rather than "canAccess"?

dwrensha replied:
Stringified because as far as I understand, keys in plain javascript objects are always strings. I considered using an `Array` rather than a plain object here, but that seemed weird.

-1 because that makes it somewhat easier to iterate through the permissions and to find the maximum key.

kentonv replied:
> keys in plain javascript objects are always strings

Sort of but not exactly. An array in Javascript is actually an object:

    > var x = ["foo", "bar"]
    > typeof x
    'object'
    > Object.keys(x)
    [ '0', '1' ]
    > x.baz = "qux"
    > x
    [ 'foo', 'bar', baz: 'qux' ]
    > Object.keys(x)
    [ '0', '1', 'baz' ]

Similarly, objects can have numeric keys:

    > ({0: "foo", 1: "bar"})[1]
    'bar'

Theoretically a non-string key is converted to a string for all purposes, but that's not actually what happens under the hood. v8 treats numeric keys (including string keys that happen to be canonical representations of integers) differently from strings and optimizes for them. Internally, the difference between an object and an array is actually pretty fuzzy.

But in any case, it is perfectly fine to use numbers as keys, or use an array with extra non-numeric fields.

> -1 because that makes it somewhat easier to iterate through the permissions and to find the maximum key.

But it's significantly less self-documenting, and if we end up with other pseudo-permissions in the future it gets uglier...

dwrensha commented on shell/packages/sandstorm-permissions/permissions.js:
```diff
@@ -315,65 +314,60 @@ class Variable {
+    // A map from permission ID to boolean. If a permission ID is present in this map,
+    // then this token's roleAssignment includes that permission. If the corresponding
+    // value is `true`, then we've proven that this token receives this permission and
+    // therefore propagates it.
```
"propagates it if the requirements are met", I should probably say.

kentonv commented on shell/packages/sandstorm-permissions/permissions.js:
```diff
@@ -399,136 +433,447 @@ class Context {
+              variable.requirementsTailList.push(tokenId);
+            }
+          });
+        }
```
Hmm, what about other types of requirements? `tokenValid` and `userIsAdmin`?

dwrensha replied:
I'm actually not sure that I understand the meaning of `tokenValid` requirements. Are they satisfied if the token merely exists and hasn't been revoked? Or (assuming the token is a UiView token) do they also require receives some set of permissions, similar to the `permissionsHeld` requirement? In the latter case, we would need to add a field, I think.

kentonv replied:
A `tokenValid` requirement comes about when a capability was transmitted across a non-UiView capability. That is, the `tokenValid` requirement should always point to a non-UiView token. The requirement passes if restoring that token would currently succeed. This requires checking that the token and its ancestors (parentToken chain) all exist and are not revoked, and that all of their respective requirements are satisfied.

kentonv commented on shell/packages/sandstorm-permissions/permissions.js:
```diff
@@ -399,136 +433,447 @@ class Context {
+    return true;
+  }
+
+  _ensureVariableExists(grainId, vertexId, permissionId) {
```
I think you can collapse _ensureVariableExists() into getVariable() at this point and save some redundant operations. Callers can call getVariable() and ignore the result if they just wanted to make sure the variable exists...

kentonv commented on shell/packages/sandstorm-permissions/permissions.js:
```diff
@@ -399,136 +433,447 @@ class Context {
-    check(grainId, String);
-    const grain = db.getGrain(grainId);
-    this.grains[grainId] = grain;
+  addGrains(db, grainIds) {
```
It looks like various callers of this method could conceivably pass in an empty grainIds, in which case the database query is a waste.

kentonv commented on shell/packages/sandstorm-permissions/permissions.js:
```diff
@@ -399,136 +433,447 @@ class Context {
+      this.addGrains(db, [grainId]);
+    }
+
+    this.activateRelevantTokens(grainId, vertexId, permissionSet);
```
This method appears to take only two arguments, but you're passing a third here.

kentonv replied:
Incidentally it seems useful for it to take permissionSet as an argument, too, especially when checking requirements. E.g. if we have a requirement that Bob has write access then we might waste time following tokens that only provide read access. But it also seems like this would be more complicated to implement since now you need partially-activated tokens, or something.

kentonv commented on shell/packages/sandstorm-permissions/permissions.js:
```diff
@@ -399,136 +433,447 @@ class Context {
+      if (tokenId) {
+        const token = this.tokensById[tokenId];
+
+        let sharerId = (token.parentToken && "t:" + token.parentToken) || ("i:" + token.identityId);
```
`(a && b) || c` => `a ? b : c`?

kentonv commented:
OK I did the commentary thing again. See/merge: https://github.com/sandstorm-io/sandstorm/commit/4dddb34eedbdd8cca0ed2b11620fcf136615c3a5

kentonv commented:
Note: Github apparently buggily claims that comments were on an "outdated diff" when really they were on a diff that was too large for Github to display by default.

dwrensha commented:
Updated to handle `tokenValid` and `userIsAdmin` requirements. I also updated `checkRequirements()` in core.js, which was not validating parent tokens.

dwrensha commented:
rebasing to resolve conflicts...
